### PR TITLE
test(DRIVERS-2384): test deployed faas functions

### PR DESCRIPTION
**Pull request by durran**
Node PR: https://github.com/mongodb/node-mongodb-native/pull/3569
Tools PR: https://github.com/mongodb-labs/drivers-evergreen-tools/pull/274

Please complete the following before merging:

- [ ] Update changelog.
- [ ] Make sure there are generated JSON files from the YAML test files.
- [x] Test changes in at least one language driver.
- [ ] Test these changes against all server versions and topologies (including standalone, replica set, sharded clusters, and serverless).

### CODE REVIEW

File: source/faas-automated-testing/faas-automated-testing.rst [hunk 1]
```diff
@@ -0,0 +1,299 @@
...
+
+.. code:: none
+
+  #!/bin/bash
+  set -o errexit  # Exit the script with error if any of the commands fail
+  . ${DRIVERS_TOOLS}/.evergreen/run-deployed-lambda-aws-tests.sh
+
+Description of the behaviour of run-deployed-lambda-aws-tests.sh:
+
+- Creates a new Atlas cluster in a specific driver project
```
Comment (ShaneHarvey) [inline source/faas-automated-testing/faas-automated-testing.rst:289]: Do we expect to run into any rate limits for cluster creation like we did during the astrolabe testing project?

Comment (durran) [reply source/faas-automated-testing/faas-automated-testing.rst:293]: Added this section: https://github.com/mongodb/specifications/pull/1387/commits/f1ff81b7106843b6314c4f4959d24702732a352d

File: source/faas-automated-testing/faas-automated-testing.rst [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+
+Prerequisites
+`````````````
+
+For the initial local setup the following are required:
+
+- The docker daemon running on the local machine.
+- The [https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/install-sam-cli.html](AWS SAM CLI)
+
+The following environment variables MUST be present (See DRIVERS-2384 google doc for values):
```
Comment (alcaeus) [inline source/faas-automated-testing/faas-automated-testing.rst:57]:
```suggestion
The following environment variables MUST be present:
```

Details about where credentials are available should not be contained in the spec, as this information might change in the future and shouldn't require a spec change. The environment variables here are also required for a number of other tests, so there's a good chance drivers will already have them set in their CI environment.

File: source/faas-automated-testing/faas-automated-testing.rst [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+
+- `AWS_ACCESS_KEY_ID`
+- `AWS_SECRET_ACCESS_KEY`
+- `AWS_REGION` - Set to us-east-1
+- `MONGODB_URI` - The local MongoDB instance
+
+Project Initialization
+``````````````````````
+
+Create the new project via SAM and follow the prompts:
```
Comment (alcaeus) [inline source/faas-automated-testing/faas-automated-testing.rst:67]: Does the binary support command line options that we can leverage to reduce the number of prompts? Looking at the prompts below, the only prompt that drivers will have to answer differently is that of the runtime, the rest could be hardcoded in the command.

Disregard this comment ff the command does not support options to skip some prompts.

File: source/faas-automated-testing/faas-automated-testing.rst (hunk 2 above)
Comment (durran) [reply source/faas-automated-testing/faas-automated-testing.rst]: I've removed the wording now.

File: source/faas-automated-testing/faas-automated-testing.rst (hunk 3 above)
Comment (durran) [reply source/faas-automated-testing/faas-automated-testing.rst:67]: There is a `--no-interactive` option that can be used, but it does require creating a template config for every single language that needs to be supplied as a `--location` option. I wasn't thinking that would be part of this as I wanted each language to have control over the setup and not the spec. I'm happy though to do that if you think it's beneficial but will probably still require someone from each driver team to review those templates and make the necessary adjustments.

**Review by alcaeus — APPROVED** [review]

**Review by ShaneHarvey — APPROVED** [review]

File: source/faas-automated-testing/faas-automated-testing.rst [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+
+Change default timeout to 30 seconds:
+
+.. code:: yaml
+
+  Globals:
+    Function:
+      Timeout: 30
+
+Add a root parameter for for the MongoDB connection string:
```
Comment (JamesKovacs) [inline source/faas-automated-testing/faas-automated-testing.rst:161]: for for => for

File: source/faas-automated-testing/faas-automated-testing.rst [hunk 5]
```diff
@@ -0,0 +1,303 @@
...
+
+.. code:: yaml
+
+  Parameters:
+    MongoDbUri:
+      Type: String
+      Description: The MongoDB connection string.
+
+Replace all instances in the template yaml of `HelloWorld` with `MongoDB` and then
```
Comment (JamesKovacs) [inline source/faas-automated-testing/faas-automated-testing.rst:170]: template.yaml

File: source/faas-automated-testing/faas-automated-testing.rst [hunk 6]
```diff
@@ -0,0 +1,303 @@
...
+
+  sam build
+  sam local invoke --parameter-overrides "MongoDbUri=${MONGODB_URI}"
+
+
+Implementing the Function
+`````````````````````````
+
+Drivers MUST setup the function as would be done in their appropriate language. In
+the function implementation driver MUST:
```
Comment (JamesKovacs) [inline source/faas-automated-testing/faas-automated-testing.rst:198]: driver MUST => the driver MUST

File: source/faas-automated-testing/faas-automated-testing.rst [hunk 7]
```diff
@@ -0,0 +1,303 @@
...
+        DRIVERS_ATLAS_GROUP_ID: ${DRIVERS_ATLAS_GROUP_ID}
+        LAMBDA_STACK_NAME: dbx-node-lambda
+        AWS_REGION: us-east-1
+        AWS_ACCESS_KEY_ID: ${AWS_ACCESS_KEY_ID}
+        AWS_SECRET_ACCESS_KEY: ${AWS_SECRET_ACCESS_KEY}
+        AWS_SESSION_TOKEN: ${AWS_SESSION_TOKEN}
+
+Drivers MUST run the function on a single variant in Evergreen, in order to not
+potentially hit the Atlas API rate limit. The variant itself MUST be either a
+RHEL7/8 or Ubuntu 18/20 variant in order to have the SAM CLI installed.
```
Comment (JamesKovacs) [inline source/faas-automated-testing/faas-automated-testing.rst:281]: Ubuntu 18 is EOL and is being retired from EG. We should recommend RHEL7/8 or Ubuntu 20.

**Review by JamesKovacs — CHANGES_REQUESTED** [review]
Fixes to minor typos. Otherwise LGTM.

File: source/faas-automated-testing/faas-automated-testing.rst (hunk 7 above)
Comment (durran) [reply source/faas-automated-testing/faas-automated-testing.rst]: Updated.

File: source/faas-automated-testing/faas-automated-testing.rst (hunk 6 above)
Comment (durran) [reply source/faas-automated-testing/faas-automated-testing.rst]: Updated.

File: source/faas-automated-testing/faas-automated-testing.rst (hunk 5 above)
Comment (durran) [reply source/faas-automated-testing/faas-automated-testing.rst]: Updated.

File: source/faas-automated-testing/faas-automated-testing.rst (hunk 4 above)
Comment (durran) [reply source/faas-automated-testing/faas-automated-testing.rst]: Updated.

**Review by JamesKovacs — APPROVED** [review]
LGTM

**Pull request merged**